framework/sandstone.cpp

```diff
@@ -1391,79 +1381,142 @@ static ChildExitStatus wait_for_child(int ffd, intptr_t child, int *tc, const st
                 enable_interrupt_catch();       // re-arm SIGINT handler
             } else {
                 // for any other signal (e.g., SIGTERM), we don't
-                break;
+                return int(signal);
             }
         }
-    }
+        if (ret <= 0)
+            return 0;
 
-    if (auto [sig, count] = last_signal(); __builtin_expect(sig, 0)) {
-        // we caught a SIGINT
-        // child has likely not been able to write results
-        logging_print_results({ TestResult::Interrupted }, tc, test);
-        logging_printf(LOG_LEVEL_QUIET, "exit: interrupted\n");
-        logging_flush();
-
-        // now exit with the same signal
-        disable_interrupt_catch();
-        raise(sig);
-        _exit(128 | sig);           // just in case
-    } else if (__builtin_expect(exit_status == TimedOut, false)) {
-        /* timed out, force the child to exit */
-        debug_hung_child(child);
-        terminate_child();
-        if (!do_wait(20s)) {
-            /* timed out again, kill the process */
-            kill_child();
-            if (!do_wait(20s)) {
-                log_platform_message(SANDSTONE_LOG_ERROR "# Child %td is hung and won't exit even after SIGKILL",
-                                     child);
-            }
+        if (pollfd &pfd = children.pollfds.back(); pfd.revents & POLLIN) {
+            // one child (or more than one) is crashing
+            debug_crashed_child();
         }
 
-        return { TestResult::TimedOut };
-    }
-    return test_result_from_exit_code(info);
+        // check if any of the children have exited
+        for (int i = 0; i < children.pollfds.size() - 1; ++i) {
+            pollfd &pfd = children.pollfds[i];
+            if (pfd.revents == 0)
+                continue;
+
+            // wait for this child
+            struct forkfd_info info;
+            EINTR_LOOP(ret, forkfd_wait4(pfd.fd, &info, WNOHANG, nullptr));
+            if (ret == -1) {
+                if (errno == EAGAIN)
+                    continue;           // shouldn't happen...
+                perror("forkfd_wait");
+                exit(EX_OSERR);
+            }
+            forkfd_close(pfd.fd);
+            pfd.fd = -1;
+            pfd.events = 0;
+            children.handles[i] = 0;
+            children.results[i] = test_result_from_exit_code(info);
+            --children_left;
+        }
+        return 0;
+    };
+    auto kill_children = [&] {
+        for (pid_t child : children.handles) {
+            if (child)
+                kill(child, SIGKILL);
+        }
+    };
 #elif defined(_WIN32)
-    (void) ffd;
+    auto kill_children = [] { };
+    auto single_wait = [&](milliseconds timeout) {
+        bool bWaitAll = false;
+        auto handles = reinterpret_cast<const HANDLE *>(children.handles.data());
+        DWORD result = WaitForMultipleObjects(children.handles.size(), handles,
```

Comment:
[`WaitForMultipleObjects`](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitformultipleobjects) is limited to `MAXIMUM_WAIT_OBJECTS` objects, which is 64.

That's enough for 2048 cores (16 sockets of 128 or 8 sockets of 256 being the closest, worst-case scenario); after that we'll have a problem.